malsim/sims/mal_simulator.py

```diff
@@ -75,1173 +45,375 @@ def __init__(
+        all_attackers_terminated = True
 
         # Peform agent actions
-        for agent in self.agents:
-            action = actions[agent]
-            if action[0] == 0:
-                # Agent wants to wait - do nothing
-                continue
-
-            action_step = action[1]
-
-            if self.agents_dict[agent]["type"] == "attacker":
-                performed_actions[agent] = \
-                    self._attacker_step(agent, action_step)
-
-            elif self.agents_dict[agent]["type"] == "defender":
-                defender_actions, prevented_attack_steps = \
-                    self._defender_step(agent, action_step)
-                performed_actions[agent] = defender_actions
+        # Note: by design, defenders perform actions
+        # before attackers (see _register_agent)
+        for agent_name in self.agents:
+            agent = self._agents_dict[agent_name]
+            agent_actions = actions.get(agent_name, [])
+            match agent.type:
+
+                case AgentType.ATTACKER:
+                    enabled = self._attacker_step(agent, agent_actions)
+                    enabled_nodes += enabled
+                    if not agent.terminated:
+                        all_attackers_terminated = False
+
+                case AgentType.DEFENDER:
+                    enabled, disabled = \
+                        self._defender_step(agent, agent_actions)
+                    enabled_nodes += enabled
+                    disabled_nodes += disabled
+
+                case _:
+                    logger.error(
+                        'Agent %s has unknown type: %s',
+                        agent.name, agent.type
+                    )
 
-            else:
-                logger.error(
-                    'Agent %s has unknown type: %s',
-                    agent, self.agents_dict[agent]["type"])
-
-        observations, rewards, terminations, truncations, infos = (
-            self._observe_and_reward(
-                performed_actions,
-                prevented_attack_steps
-            )
-        )
+        if all_attackers_terminated:
+            # Terminate all agents if all attackers are terminated
+            logger.info("All attackers are terminated")
+            for agent in self._agents_dict.values():
+                agent.terminated = True
+
+        if self.cur_iter >= self.max_iter:
+            # Truncate all agents when max iter is reached
+            logger.info("Max iteration reached - all agents truncated")
+            for agent in self._agents_dict.values():
+                agent.truncated = True
+
+        agents_to_remove = set()
+        for agent_name in self.agents:
+            agent = self._agents_dict[agent_name]
+            if agent.terminated or agent.truncated:
+                logger.info("Removing agent %s", agent.name)
+                agents_to_remove.add(agent.name)
+
+        for agent in agents_to_remove:
+            self.agents.remove(agent)
```

Comment:
```suggestion
        agents_to_remove = set()
        for agent_name in self.agents:
            agent = self._agents_dict[agent_name]
            if agent.terminated or agent.truncated:
                logger.info("Removing agent %s", agent.name)
                self.agents.remove(agent)
```

This should work the same right?

Reply:
Hm, do you mean we can be removing items from the list we are iterating over? I don't think that would work.

Reply:
Ah right. I updated the suggestion to loop over a copy of the list.

Reply:
Yes, I like that, since it is just a list of strings and not expected to be very large this works well and is fewer lines of code. Will implement.